Move CIB transaction management to client and fix controller regression

nrwahl2 opened the pull request:
This fixes a regression in the controller join sequence introduced by https://github.com/ClusterLabs/pacemaker/commit/5e3b3d142cc41376e8a1a88faa12d6062e726035. As of that commit, there are multiple asynchronous calls when building each transaction in do_dc_join_ack(). If multiple nodes were joining in parallel, the DC may try to initiate a transaction for one node while a transaction was already open for another node. However, the DC's CIB client can have only one active transaction at a time.

Now, the transaction is created, sent to the CIB manager, and freed synchronously. There is no opportunity for another node's join ack to try to initiate a transaction in the meantime.

Fixes [T690](https://projects.clusterlabs.org/T690)

nrwahl2 commented:
@kgaillot Here's a sneak peak since you're about to be out for the weekend. I still need to test, fix any bugs that I find, and update the controller code to use the new transaction interface (i.e., no callbacks for init/extend).

nrwahl2 commented on include/crm/cib/internal.h:
```diff
@@ -70,6 +70,7 @@
 /* notify sub-types */
 #  define T_CIB_PRE_NOTIFY	"cib_pre_notify"
 #  define T_CIB_POST_NOTIFY	"cib_post_notify"
+#  define T_CIB_TRANSACTION		"cib_transaction"
```
whoops

kgaillot commented on lib/cib/cib_client.c:
```diff
@@ -390,23 +390,66 @@ cib_client_erase(cib_t * cib, xmlNode ** output_data, int call_options)
+        rc = pcmk_rc_already;
+    }
+
+    cib->transaction = create_xml_node(NULL, T_CIB_TRANSACTION);
```
I assume you want to skip this if rc != pcmk_rc_ok

nrwahl2 replied:
Thanks

kgaillot commented on daemons/based/based_transaction.c:
```diff
@@ -14,270 +14,59 @@
+        CRM_CHECK((operation != NULL)
+                  && pcmk_is_set(operation->flags, cib__op_attr_transaction)
+                  && (host == NULL),
+                  rc = EOPNOTSUPP);
```
The CIB is proxied to Pacemaker Remote nodes, so we can't assume the client and server came from the same Pacemaker version

nrwahl2 replied:
How are we assuming that here?

Transactions haven't made it into a release yet, luckily. I'm only assuming the client isn't running a version that uses the experimental server-based implementation of transactions. I think that's reasonable since it was only in the main branch and we quickly introduced a controller regression based on it.

kgaillot replied:
Future-proofing. We can't assume (for all future combinations of client and server versions) that the client has rejected an unsupported request. The client may be running an older or newer version with a different conception of supported than the server.

Not a big deal since we're re-validating here, but probably doesn't merit a CHECK

kgaillot reviewed (commented):
much cleaner implementation

nrwahl2 commented:
Flipped to ready for review. I still haven't run CTSlab because I can't get it to work. Details are in our Slack channel.

nrwahl2 commented:
Still getting the `error: Input I_STOP received in state S_FINALIZE_JOIN from notify_crmd` messages periodically despite the commit that ignores CIB replace notifications if we're shutting down

kgaillot commented on include/crm/cib/cib_types.h:
```diff
@@ -282,7 +282,7 @@ typedef struct cib_api_operations_s {
      * \return Legacy Pacemaker return code
      */
-    int (*init_transaction)(cib_t *cib, int call_options);
+    int (*init_transaction)(cib_t *cib);
```
also from doxygen

nrwahl2 replied:
Thanks

kgaillot reviewed (commented):
Through "Drop call options" looks good

nrwahl2 commented on include/crm/common/nvpair.h:
```diff
@@ -46,7 +46,6 @@ void hash2smartfield(gpointer key, gpointer value, gpointer user_data);
 GHashTable *xml2list(const xmlNode *parent);
 
 const char *crm_xml_add(xmlNode *node, const char *name, const char *value);
-const char *crm_xml_replace(xmlNode *node, const char *name, const char *value);
```
Whoops

nrwahl2 commented:
100 rounds of `cts-lab` ran mostly fine through "Drop call options." There were a fair number of "Local CIB differs" messages, especially during "RestartOnebyOne"... possibly more than usual but hard to say.

Ignore the two failures due to failure to start `pacemaker_remote`. Somehow node 44 had an empty systemd unit file for `pacemaker_remote.service`.

It's worth noting there were also two instances of the below, which I haven't seen before. They didn't result in test failures, though I'm not sure how a subdaemon PID got out of whack twice.
```
Aug 07 01:10:18	BadNews: Aug 07 01:07:07 fastvm-rhel-9-0-43 pacemakerd[202278]:  error: Could not get executable for PID 202281: No such file or directory
Aug 07 02:19:57	BadNews: Aug 07 02:19:22 fastvm-rhel-9-0-44 pacemakerd[3228]:  error: Could not get executable for PID 3233: No such file or directory
```

https://gist.github.com/nrwahl2/8e164961a55073b37eddf31db3293269

nrwahl2 commented:
Updated:
* Addressed review
* Updated the `configure.ac` fix in the first commit
* Drop the `attrd_write_attribute()` commit and the ones that followed it

nrwahl2 commented:
Can retest after you look at the cts-lab output comment

kgaillot commented:
> It's worth noting there were also two instances of the below, which I haven't seen before. They didn't result in test failures, though I'm not sure how a subdaemon PID got out of whack twice.
> 
> ```
> Aug 07 01:10:18	BadNews: Aug 07 01:07:07 fastvm-rhel-9-0-43 pacemakerd[202278]:  error: Could not get executable for PID 202281: No such file or directory
> Aug 07 02:19:57	BadNews: Aug 07 02:19:22 fastvm-rhel-9-0-44 pacemakerd[3228]:  error: Could not get executable for PID 3233: No such file or directory
> ```

This shows up now and then (previously as "Could not read from /proc/$PID/exe") despite the "Check again to filter out races" block above the code. It's harmless though annoying.

nrwahl2 commented on lib/cib/cib_file.c:
```diff
@@ -348,11 +348,10 @@ cib_file_perform_op_delegate(cib_t *cib, const char *op, const char *host,
 
-    request = cib_create_op(cib->call_id, op, host, section, data, call_options,
-                            user_name);
+    request = cib__create_op(cib, op, host, section, data, call_options,
```
This still doesn't get NULL-checked as of this commit but it's fixed as of the next commit... not worried about it now

nrwahl2 commented on lib/cib/cib_utils.c:
```diff
@@ -485,13 +485,26 @@ cib_perform_op(const char *op, int call_options, cib__op_fn_t fn, bool is_query,
-                crm_xml_replace(scratch, XML_ATTR_UPDATE_CLIENT,
-                                crm_element_value(req, F_CIB_CLIENTNAME));
-                crm_xml_replace(scratch, XML_ATTR_UPDATE_USER, crm_element_value(req, F_CIB_USER));
+                crm_xml_add(scratch, XML_ATTR_UPDATE_ORIG, origin);
```
Since `CRM_LOG_ASSERT()` doesn't keep us from proceeding, we should still NULL-check `origin` to decide whether to add or remove it. I'll make a note for a future PR.

kgaillot replied:
Yep, and we can drop the assert -- it was probably only useful when this code was first being developed, and nothing horrible happens if it's NULL

nrwahl2 commented on daemons/based/based_callbacks.c:
```diff
@@ -282,24 +281,10 @@ cib_common_callback_worker(uint32_t id, uint32_t flags, xmlNode * op_request,
+    /* Requests with cib_transaction set should not be sent to based directly
+     * (outside of a commit-transaction request)
+     */
+    CRM_CHECK(!pcmk_is_set(call_options, cib_transaction), return);
```
Doesn't warrant a check

nrwahl2 commented:
Rebased on current main and addressed my own minor review points.
